src/control/cmd/dmg/main.go

```diff
@@ -82,37 +79,89 @@ func appSetup(broadcast bool) error {
 	return nil
 }
 
-func main() {
-	if dmgMain() != nil {
-		os.Exit(1)
-	}
+type broadcastCmd struct{}
+
+func (broadcastCmd) isBroadcast() {}
+
+type broadcaster interface {
+	isBroadcast()
 }
 
-func dmgMain() error {
-	// Set default global logger for application.
-	log.NewDefaultLogger(log.Debug, "", os.Stderr)
+type connectedCmd struct {
+	conns client.Connect
+}
+
+func (cmd *connectedCmd) setConns(conns client.Connect) {
+	cmd.conns = conns
+}
+
+type connector interface {
+	setConns(client.Connect)
+}
+
+func exitWithError(err error) {
+	log.Errorf("%s exiting with error: %v", path.Base(os.Args[0]), err)
+	os.Exit(1)
+}
+
+func parseOpts(args []string) (*cliOptions, client.Connect, error) {
+	opts := new(cliOptions)
+	conns := client.NewConnect()
 
-	// Parse cli args and either execute subcommand then exit or
-	// drop into shell if no subcommand is specified.
 	p := flags.NewParser(opts, flags.Default)
 	p.SubcommandsOptional = true
+	p.CommandHandler = func(cmd flags.Commander, args []string) error {
+		if cmd == nil {
+			return nil
+		}
+
+		_, shouldBroadcast := cmd.(broadcaster)
+		if err := appSetup(shouldBroadcast, opts, conns); err != nil {
+			return err
+		}
+		if wantsConn, ok := cmd.(connector); ok {
+			wantsConn.setConns(conns)
+		}
+		if err := cmd.Execute(args); err != nil {
+			return err
+		}
+
+		// Stop execution here and exit normally
+		// after the command is executed successfully.
+		os.Exit(0)
+		return nil
+	}
 
-	_, err := p.Parse()
+	unparsed, err := p.ParseArgs(args)
 	if err != nil {
-		return err
+		return nil, nil, err
+	}
+	if len(unparsed) > 0 {
+		log.Debugf("Unparsed arguments: %v", unparsed)
```

Comment:
It could. Should it? Right now it appears to just silently drop into the shell when it encountered arguments it doesn't know about. It would probably be better to fail on unknown arguments so that external scripts do the right thing.

Reply:
yes that's better, was just checking we still print the usage message when invalid arguments are given